auth-manifest/gen/src/generator.rs

```diff
@@ -0,0 +1,166 @@
+
+/// Authorization Manifest generator
+pub struct AuthManifestGenerator<Crypto: ImageGeneratorCrypto> {
+    crypto: Crypto,
+}
+
+impl<Crypto: ImageGeneratorCrypto> AuthManifestGenerator<Crypto> {
+    /// Create an instance `AuthManifestGenerator`
+    pub fn new(crypto: Crypto) -> Self {
+        Self { crypto }
+    }
+
+    pub fn generate(
+        &self,
+        config: &AuthManifestGeneratorConfig,
+    ) -> anyhow::Result<AuthorizationManifest> {
+        let mut auth_manifest = AuthorizationManifest::default();
+
+        if config.image_metadata_list.len() > AUTH_MANIFEST_IMAGE_METADATA_MAX_COUNT {
+            eprintln!(
+                "Encountered an error converting image_metadata_list, only {} entries supported.",
+                AUTH_MANIFEST_IMAGE_METADATA_MAX_COUNT
+            );
+            return Err(anyhow::anyhow!("Error converting image_metadata_list"));
+        }
+
+        // Generate the Image Metadata List.
+        let slice = config.image_metadata_list.as_slice();
+        auth_manifest.image_metadata_col.image_metadata_array[..slice.len()].copy_from_slice(slice);
+
+        auth_manifest.image_metadata_col.header.entry_count =
+            config.image_metadata_list.len() as u32;
+        auth_manifest.image_metadata_col.header.revision = 0; // [TODO] Need to update this.
+
+        // Generate the preamble.
+        auth_manifest.preamble.marker = AUTH_MANIFEST_MARKER;
+        auth_manifest.preamble.size = size_of::<AuthManifestPreamble>() as u32;
+        auth_manifest.preamble.version = config.version;
+        auth_manifest.preamble.flags = config.flags;
+
+        // Sign the vendor manifest keys.
+        auth_manifest.preamble.vendor_pub_keys = config.vendor_man_key_info.pub_keys;
+
+        let range = AuthManifestPreamble::vendor_signed_data_range();
+
+        let data = auth_manifest
+            .preamble
+            .as_bytes()
+            .get(range.start as usize..)
+            .ok_or_else(|| anyhow::anyhow!("Failed to get vendor signed data range start"))?
+            .get(..range.len())
+            .ok_or(anyhow::anyhow!(
+                "Failed to get vendor signed data range length"
+            ))?;
+
+        let digest = self.crypto.sha384_digest(data)?;
+
+        if let Some(priv_keys) = config.vendor_fw_key_info.priv_keys {
+            let sig = self.crypto.ecdsa384_sign(
+                &digest,
+                &priv_keys.ecc_priv_key,
+                &config.vendor_fw_key_info.pub_keys.ecc_pub_key,
+            )?;
+            auth_manifest.preamble.vendor_pub_keys_signatures.ecc_sig = sig;
+
+            let lms_sig = self.crypto.lms_sign(&digest, &priv_keys.lms_priv_key)?;
+            auth_manifest.preamble.vendor_pub_keys_signatures.lms_sig = lms_sig;
+        }
+
+        // Sign the owner manifest keys.
+        if let Some(owner_fw_config) = &config.owner_fw_key_info {
+            if let Some(owner_man_config) = &config.owner_man_key_info {
```

Comment:
nit: I think you can do something like this to avoid the nested ifs
```suggestion
        if let (Some(owner_fw_config), Some(owner_man_config)) = (&config.owner_fw_key_info, &config.owner_man_key_info) {
```

Reply:
Updated.